### PR TITLE
eglfs: Allow scaling down the window surface

### PR DESCRIPTION
**Pull request by giucam**
This patch introduces the QT_QPA_EGLFS_SCALE environment variable.
When set to an integer number > 1 it will scale the draws window
content down by that scale factor.

### CODE REVIEW

File: src/plugins/platforms/eglfs/qeglfscontext.cpp [hunk 1]
```diff
@@ -48,15 +48,212 @@
...
+      m_useNativeDefaultFbo(false),
+      m_fbo(0),
+      m_blitter(0)
+{
+    m_scale = qgetenv("QT_QPA_EGLFS_SCALE").toInt();
+}
+
+QEglFSContext::~QEglFSContext()
 {
+    delete m_blitter;
```
Comment (adenexter) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp:256]: delete m_fbo as well?

Or possibly use QScopedPointer for m_blitter and m_fbo.

**Comment by adenexter** [comment]
The scaling works correctly for me, but I'm seeing a couple of scene graph items that aren't rendering when a scale is specified.

Specifically the glass effect which may be because the fragment shader samples screen coordinates.  And the launcher background when it is partially open, which is a little harder to explain.

File: src/plugins/platforms/eglfs/qeglfscontext.cpp [hunk 2]
```diff
@@ -48,15 +48,212 @@
...
+class Blitter : public QOpenGLFunctions
+{
+public:
+    Blitter(QEglFSContext *context)
+        : m_context(context)
+    {
+        initializeOpenGLFunctions();
+        m_blitProgram = new QOpenGLShaderProgram();
+        m_blitProgram->addShaderFromSourceCode(QOpenGLShader::Vertex, "attribute vec4 position;\n\
+                                                                    attribute vec4 texCoords;\n\
```
Comment (sletta) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp]: vec2

File: src/plugins/platforms/eglfs/qeglfscontext.cpp [hunk 3]
```diff
@@ -48,15 +48,212 @@
...
+                                                                    varying vec2 outTexCoords;\n\
+                                                                    void main()\n\
+                                                                    {\n\
+                                                                        gl_Position = position;\n\
+                                                                        outTexCoords = texCoords.xy;\n\
```
Comment (sletta) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp]: if texCoords is a vec2, then you can use only texCoords, without .xy

File: src/plugins/platforms/eglfs/qeglfscontext.cpp [hunk 4]
```diff
@@ -48,15 +48,212 @@
...
+            glFuncs.glBindBuffer(GL_ARRAY_BUFFER, m_vertexAttribs[i].arrayBuffer);
+            glFuncs.glVertexAttribPointer(i, m_vertexAttribs[i].size, m_vertexAttribs[i].type,
+                                          m_vertexAttribs[i].normalized, m_vertexAttribs[i].stride,
+                                          m_vertexAttribs[i].pointer);
+            glFuncs.glBindBuffer(GL_ARRAY_BUFFER, prevBuf);
+        }
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, m_minFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, m_magFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, m_wrapS);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, m_wrapT);
```
Comment (sletta) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp:212]: You didn't strictly touch the wrap modes

Comment (giucam) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp:212]: I don't see what you mean...

**Comment by giucam** [comment]
I've noticed the rendering problems, but i thought they probably are mesa bugs... iirc the emulator uses mesa 9.2 which is pretty old.
I wouldn't know how to explain them otherwise.

**Comment by adenexter** [comment]
It could be mesa, I certainly don't have a better explanation.

File: src/plugins/platforms/eglfs/qeglfscontext.cpp (hunk 4 above)
Comment (sletta) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp:212]: In the blit function, you are only setting MIN and MAG filter. You donæt touch the wrap modes, so you don't need to restore them either.

Comment (giucam) [inline src/plugins/platforms/eglfs/qeglfscontext.cpp:212]: Ah, right.

**Comment by sletta** [comment]
LGTM

**Pull request merged**